### ISSUE TITLE
`balanceTest` fails with missingness in the covariates

### ISSUE THREAD

**Issue by markmfredrickson** [markmfredrickson/RItools#92]
While working on issue #84 I noticed the following bug:

    set.seed(20121119)
    df <- data.frame(Z = rep(c(0,1), 10),
                     X1 = rnorm(20, mean = Z*2),
                     X2 = rnorm(20, mean = Z*3))
    
    df$X1[1:3] <- NA
    
    bt <- balanceTest(Z ~ X1 + X2, data = df)
    
    => Error in Covs * Uweights (from Design.R#909) : non-conformable arrays

If one uses only `X1`, there is no issue.

**Comment by benthestatistician** [comment]
Is it just that `Uweights` has the wrong dimension?  Perhaps 

1. the dimension attribute of `alignedcovs@NotMissing` got dropped prior to passing of material to `alignedToInferentials()`? Or maybe...
2.  Inside of `alignedToInferentials()`, the dimension attribute of `alignedcovs@NotMissing` is not being properly transferred to `Uweights`?

Looking at this code block in the master branch, it looks like there's square-bracket subsetting of matrices happening w/o the `, drop=FALSE` proviso...  

(Thanks for your help cleaning up after me, @markmfredrickson )

**Comment by markmfredrickson** [comment]
This response mostly for note taking. It might suggest something. Results are from debugging the example above.

First, let's look at some of the data.

```
Browse[5]> Covs
                X1         X2  (X1) (_non-null record_)
 [1,]  0.000000000  0.8961609 -0.85                   0
 [2,]  0.000000000 -2.3889877 -0.85                   0
 [3,]  0.000000000 -1.1794148 -0.85                   0
 [4,] -0.353139011  1.0385596  0.15                   0
 [5,] -0.357292659 -0.9763670  0.15                   0
...

Browse[5]> alignedcovs@NotMissing

      _non-null record_       X1
 [1,]                 1 0.000000
 [2,]                 1 0.000000
 [3,]                 1 0.000000
 [4,]                 1 1.176471
 [5,]                 1 1.176471
 [6,]                 1 1.176471
...

Browse[5]> alignedcovs@NM.Covariates
[1] 2 0 1 1
```

The `Uweights` object when created has the form:

```
Browse[5]> Uweights
            X1 _non-null record_ _non-null record_
 [1,] 0.000000                 1                 1
 [2,] 0.000000                 1                 1
 [3,] 0.000000                 1                 1
 [4,] 1.176471                 1                 1
 [5,] 1.176471                 1                 1
```

I'm a little confused as to which of these correct. The [usage in this function](https://github.com/markmfredrickson/RItools/blob/master/R/Design.R#L909) suggests that the `NotMissing` matrix should have rows equal to the number of units, while the docs state that the `NotMissing` matrix should have rows equal to the number of covariates. 

Assuming the usage is ahead of the documentation, I think that `alignedcovariates@NM.Covariates` should be `c(2, 1, 1, 1)`, so that Uweights weights X1 by the missingness aspect but all other columns have no missingness so get full weight of 1.

Is there overlap here with issue #89?

**Comment by benthestatistician** [comment]
A couple quick comments on Mark's....



> I'm a little confused as to which of these correct. The usage in this
> function
> <https://github.com/markmfredrickson/RItools/blob/master/R/Design.R#L909>
> suggests that the NotMissing matrix should have rows equal to the number
> of units, while the docs state that the NotMissing matrix should have
> rows equal to the number of covariates.
>
> Assuming the usage is ahead of the documentation,
>

Yep that's a typo in the docs.


> I think that alignedcovariates@NM.Covariates should be c(2, 1, 1, 1), so
> that Uweights weights X1 by the missingness aspect but all other columns
> have no missingness so get full weight of 1.
>

This is a correct statement of what I was trying to have happen here.

>

**Comment by markmfredrickson** [comment]
I made a fix in 4f160ab8 that fixes the problem with `alignedToInferentials` to give weights of 1 to any columns without missingness. The alternative would have been to change the way the objects create `NM.Covariates`, but in looking at the inline comments and the docs, it seemed better to do the change downstream.

@benthestatistician If you can confirm this approach, I'll merge the change and close the issue.

**Comment by benthestatistician** [comment]
I can confirm this approach. Thanks, @markmfredrickson .  Commit [master 674f354] implicitly created similar mappings within `alignDesignsByStrata()`, via `max(1L, design@NM.Covariates)`.  I overlooked this point in `alignedToInferentials()` where I ought to have done the same thing.
